recirq/seniority_zero/circuits_expt/core.py

```diff
@@ -0,0 +1,430 @@
+from recirq.seniority_zero.misc import (
+    parallelize_circuits,
+    safe_concatenate_circuits,
+    safe_tetris_circuits,
+)
+from recirq.seniority_zero.scheduling import get_tqbg_groups
+
+
+def _make_lightcone(num_qubits, depth=None):
+    """Make list of gates which are hit by a given qubit's backwards light cone.
+    Assumes a brickwall ansatz
+    """
+    if depth is None:
+        depth = num_qubits // 2
+    gates_per_layer = num_qubits // 2
+    last_layer_offset = (gates_per_layer + 1) % 2
+    timelike_gates_all_qubits = []
+    for qid in range(num_qubits):
+        # Add gate that this qubit interacts with in the last layer
+        timelike_gates_this_qubit = [[((qid - last_layer_offset) % num_qubits) // 2]]
+        # Populate backwards
+        for layer_id in range(1, num_qubits // 2):
+            # All gate ids from the previous layer are included here
+            timelike_gates_this_qubit.append(list(timelike_gates_this_qubit[-1]))
+
+            # Either add the largest gid +1 or the smallest gid -1
+            offset = (layer_id + last_layer_offset) % 2
+            if offset == 0:
+                timelike_gates_this_qubit[-1].append(
+                    (timelike_gates_this_qubit[-1][-1] + 1) % gates_per_layer
+                )
+            else:
+                timelike_gates_this_qubit[-1].insert(
+                    0, (timelike_gates_this_qubit[-1][0] - 1) % gates_per_layer
+                )
+        timelike_gates_all_qubits.append(timelike_gates_this_qubit)
+    return timelike_gates_all_qubits
+
+
+# Indexed: _timelike_gates[num_qubits][qid][negative_layer_id]
+_timelike_gates = {num_qubits: _make_lightcone(num_qubits) for num_qubits in [4, 6, 8, 10, 12]}
+
+
+def gs_layer(qubits: List, layer_params: List[float], offset: int, gate_indices: List[int]):
+    """Single even layer of the brick wall givens-swap circuit.
+
+    Qubits are coupled as (dashed lines represent a single givens-swap gate)
+    0-1 2-3 4-5 6-7 8-9
+    or (if odd_flag == True)
+    0 1-2 3-4 5-6 7-8 9
+    |_________________|
+
+    Args:
+        qubits [List[cirq.Qid]]: array with the indexing for the qubit
+        layer_params [List[float]]: array of length len(qubits)//2 with
+            the parameters for the givens rotations of the layer
+        odd_flag [bool]: whether this is an even or odd layer
+    """
+    assert len(layer_params) == len(qubits) // 2
+
+    circuits = [
+        gsgate(qubits[2 * i + offset], qubits[(2 * i + 1 + offset) % len(qubits)], layer_params[i])
+        for i in gate_indices
+    ]
+    circuit = parallelize_circuits(*circuits)
+
+    return circuit
+
+
+def givens_swap_network(
+    qubits: List,
+    params: List[float],
+    depth: int,
+    target_qids: Optional[List[cirq.Qid]] = None,
+    inverse: Optional[bool] = False,
+    global_echo: Optional[bool] = False,
+    echoed_layers: Optional[List] = None,
+):
```

Comment:
return type annotation and/or documentation

Reply:
I think I've gotten this